File: src/main/java/ini/cx3d/physics/interfaces/PhysicalObject.java
package ini.cx3d.physics.interfaces;

import ini.cx3d.localBiology.interfaces.CellElement;
import ini.cx3d.synapses.Excrescence;

import java.awt.*;
import java.util.AbstractSequentialList;
import java.util.Hashtable;

/**
 * Created by lukas on 09.03.16.
 */
public interface PhysicalObject extends PhysicalNode {
	@Override
	ini.cx3d.swig.NativeStringBuilder simStateToJson(ini.cx3d.swig.NativeStringBuilder sb);

	/** Returns true because this object is a PhysicalObject.*/
	boolean isAPhysicalObject();

	/**
	 * Returns true if this <code>PhysicalObject</code> and the <code>PhysicalObject</code> given as
	 * argument have a mother-daughter or sister-sister (e.g. two daughters of a same mother) relation.
	 */
	boolean isRelative(PhysicalObject po);

	/** Returns the <code>CellElement</code>linked to this <code>PhysicalObject</code>.*/
	CellElement getCellElement();

	/** Adds an <code>Excrescence</code> instance to the Excrescence list of this
	 * <code>PhysicalObject</code>.*/
	void addExcrescence(Excrescence ex);

	/** Removes an <code>Excrescence</code> instance to the Excrescence list of this
	 * <code>PhysicalObject</code>.*/
	void removeExcrescence(Excrescence ex);

	/** Active displacement of the point mass of this <code>PhysicalObject</code>. ("active" means
	 * "triggered by a biological process" like in growth or migration). This method MUST NOT be used
	 * for displacement by purely passive (physical) force.
	 * @param speed in microns/hour
	 * @param direction a vector indicating the direction of movement
	 */
	void movePointMass(double speed, double[] direction);

	/** Compute physical forces, and move accordingly to one simulation time step.*/
	void runPhysics();

	/**
	 * Returns the inter-object force that the <code>PhysicalObject</code> in which the method is called applies
	 * onto the <code>PhysicalSphere</code> given as argument.
	 * @param s
	 * @return
	 */
	double[] getForceOn(ini.cx3d.physics.interfaces.PhysicalSphere s);

	/**
	 * Returns true if this <code>PhysicalObject</code> is in contact, i.e. if it is
	 * close enough to the <code>PhysicalObject</code> given as argument.
	 * @param o
	 * @return
	 */
	boolean isInContact(PhysicalObject o);

	/**
	 * Returns all the neighboring objects considered as being in contact with this PhysicalObject.
	 * @return
	 */
	AbstractSequentialList<PhysicalObject> getPhysicalObjectsInContact();

	void changeDiameter(double speed);

	void changeVolume(double speed);

	/**
	 * Returns the position in the global coordinate system (cartesian coordinates)
	 * of a point expressed in polar coordinates (cylindrical or spherical).
	 * @param positionInPolarCoordinates a point defined in polar coordinate system of a PhysicalObject
	 * @return [x,y,z] the absolute value in space
	 */
	double[] transformCoordinatesPolarToGlobal(double[] positionInPolarCoordinates);

	/**
	 * Returns the position in the polar coordinate system (cylindrical or spherical)
	 * of a point expressed in global cartesian coordinates ([1,0,0],[0,1,0],[0,0,1]).
	 * @param positionInAbsoluteCoordinates the [x,y,z] cartesian values
	 * @return the position in local coord.
	 */
	double[] transformCoordinatesGlobalToPolar(double[] positionInAbsoluteCoordinates);

	/**
	 * Returns the position in the local coordinate system (xAxis, yXis, zAxis)
	 * of a point expressed in global cartesian coordinates ([1,0,0],[0,1,0],[0,0,1]).
	 * @param positionInGlobalCoord
	 * @return
	 */
	double[] transformCoordinatesGlobalToLocal(double[] positionInGlobalCoord);

	/**
	 * Returns the position in in global cartesian coordinates ([1,0,0],[0,1,0],[0,0,1])
	 * of a point expressed in the local coordinate system (xAxis, yXis, zAxis).
	 * @param positionInLocalCoord
	 * @return
	 */
	double[] transformCoordinatesLocalToGlobal(double[] positionInLocalCoord);

	/**
	 * Returns a unit vector, pointing out of the PhysicalObject if origin at location
	 * specified in argument.
	 * @param positionInPolarCoordinates the origin of the normal vector (local cartesian coord)
	 * @return a vector pointing "out", of unitary norm (absolute cartesian coord)
	 */
	double[] getUnitNormalVector(double[] positionInPolarCoordinates);

	/** Simply adds the argument to the vector containing all the PhysicalBonds of this
	 * PhysicalObject.*/
	void addPhysicalBond(PhysicalBond pb);

	/** Simply removes the argument from the vector containing all the PhysicalBonds of this
	 * PhysicalObject. */
	void removePhysicalBond(PhysicalBond pb);

	/** Returns true if there is a PhysicalBond that fixes me to this other PhysicalObject.*/
	boolean getHasAPhysicalBondWith(PhysicalObject po);

	/**
	 * Creates a new PhysicalBond between this PhysicalObject and the one given as argument.
	 * The newly created PhysicalBond is inserted into the physical bon's list of both objects.
	 * @param po
	 * @return
	 */
	PhysicalBond makePhysicalBondWith(PhysicalObject po);

	/**
	 * If there is a PhysicalBond between this PhysicalObject and po,
	 * it is removed (in both objects).
	 * @param po the other PhysicalObject we want to test with
	 * @param removeThemAll if true, makes multiple removals (if multiple bonds)
	 * @return true if at least one PhysicalBond was removed
	 */
	boolean removePhysicalBondWith(PhysicalObject po, boolean removeThemAll);

	/** Compute diffusion of <code>IntracellularSubstances</code> with relatives in the neuron
	 * tree structure, and perform diffusion processes according to one simulation time step.*/
	void runIntracellularDiffusion();

	/**
	 * Returns the concentration of an <code>IntracellularSubstance</code> in this
	 * PhysicalObject. If not present at all, zhe value 0 is returned.
	 * @param substanceId
	 * @return
	 */
	double getIntracellularConcentration(String substanceId);

	/** Modifies the quantity (increases or decreases) of an IntracellularSubstance.
	 * If this <code>PhysicalNode</code> already has an <code>IntracellularSubstance</code>
	 * instance corresponding to the type given as argument (with the same id), the fields
	 * quantity and concentration in it will be modified, based on a computation depending
	 * on the simulation time step and the space volume (for the latter : only if volumeDependant
	 * is true in the IntracellularSubstance). If there is no such IntracellularSubstance
	 * instance already, a new instance is requested from ECM.
	 * <p>
	 * This method is not used for diffusion, but only by biological classes...
	 * @param id the name of the Substance to change.
	 * @param quantityPerTime the rate of quantity production
	 */
	void modifyIntracellularQuantity(String id, double quantityPerTime);

	/** Returns the concentration of a membrane bound IntracellularSubstance on
	 * this PhysicalObject. Recall that by definition, the PhysicalObject are
	 * considered as expressing a cell type specific protein as well as the universal
	 * marker "U".
	 * @param id
	 * @return
	 */
	double getMembraneConcentration(String id);

	/** Modifies the quantity (increases or decreases) of an membrane-bound chemical.
	 *
	 * This method is not used for diffusion, but only by biological classes...
	 * @param id the name of the Substance to change.
	 * @param quantityPerTime the rate of quantity production
	 */

	void modifyMembraneQuantity(String id, double quantityPerTime);

	/** Returns the <code>java.awt.Color</code> used to draw this PhysicalObject in the GUI. */
	Color getColor();

	/** Sets the <code>java.awt.Color</code> used to draw this PhysicalObject in the GUI. */
	void setColor(Color color);

	/** Returns a copy of the masslocation.*/
	double[] getMassLocation();

	/**
	 * - CAUTION : Never use this method to move a PhysicalObject, because the physics is not updated.
	 * <p>
	 * - Never?
	 * <p>
	 *  - I said NEVER !
	 * @param massLocation the massLocation to set
	 */
	void setMassLocation(double[] massLocation);

	/** Returns the "Up" direction for a Sphere, and the axis direction for a Cylinder*/
	double[] getAxis();

	/** Returns the first axis of the local coordinate system.*/
	double[] getXAxis();

	/** Sets the first axis of the local coordinate system. Should have a norm of 1.*/
	void setXAxis(double[] axis);

	/** Returns the second axis of the local coordinate system.*/
	double[] getYAxis();

	/** Sets the second axis of the local coordinate system. Should have a norm of 1*/
	void setYAxis(double[] axis);

	/** Returns the third axis of the local coordinate system.*/
	double[] getZAxis();

	/** Sets the third axis of the local coordinate system. Should have a norm of 1*/
	void setZAxis(double[] axis);

	/** Only for GUI display. Total force on this objects point mass, last time it was computed.
	 * 3 first components give the x,y,z coord, and last one if movement was applied (<0 means no).*/
	double[] getTotalForceLastTimeStep();

	/** Returns true if this object still plays a role in the simulation.
	 * For instance a PhysicalObject associated with a neurite that just retracted
	 * is not exxisting. */
	boolean isStillExisting();

	/** The role of the method is to indicate that an object is about to
	 * be garbage Collected. Caution: don't use this method! */
	void setStillExisting(boolean stillExists);

	/** If true, this PhysicalObject will be run by the Scheduler on the next occasion.*/
	boolean isOnTheSchedulerListForPhysicalObjects();

	/** If true, this PhysicalObject will be run by the Scheduler on the next occasion.*/
	void setOnTheSchedulerListForPhysicalObjects(
			boolean onTheSchedulerListForPhysicalObjects);

	/** Returns the vector containing all the PhysicalBonds of this PhysicalObject.*/
	AbstractSequentialList<PhysicalBond> getPhysicalBonds();

	/** Sets the vector containing all the PhysicalBonds of this PhysicalObject.
	 * This methof should not be used during the simulation. */
//	void setPhysicalBonds(Vector<PhysicalBond> physicalBonds);

	/** Returns the vector containing all the Excrescences (PhysicalSpine, PhysicalBouton).*/
	AbstractSequentialList<Excrescence> getExcrescences();

	/** Sets the vector containing all the Excrescences (PhysicalSpine, PhysicalBouton).
	 * This method should not be used during a simulation. */
//	void setExcrescences(Vector<Excrescence> excrescences);

	/** Returns the adherence to the extracellular matrix, i.e. the static friction
	 * (the minimum force amplitude needed for triggering a movement). */
	double getAdherence();

	/** Sets the adherence to the extracellular matrix, i.e. the static friction
	 * (the minimum force amplitude needed for triggering a movement). */
	void setAdherence(double adherence);

	/** Returns the mass, i.e. the kinetic friction
	 * (scales the movement amplitude, therefore is considered as the mass).*/
	double getMass();

	/** Sets the mass, i.e. the kinetic friction
	 * (scales the movement amplitude, therefore is considered as the mass).*/
	void setMass(double mass);

	double getDiameter();

	/**
	 * Sets the diameter to a new value, and update the volume accordingly.
	 * is equivalent to setDiamater(diameter, true)
	 * @param diameter
	 */
	void setDiameter(double diameter);

	/**
	 * Sets the diameter. The volume is sets accordingly if desired.
	 *
	 * @param diameter the new diameter
	 * @param updateVolume if true, the volume is set to match the new diameter.
	 */
	void setDiameter(double diameter, boolean updateVolume);

	/** Returns the volume of this PhysicalObject.*/
	double getVolume();

	/**
	 * Sets the volume, and (optionally) recomputes an new diameter.
	 * THIS METHOD SHOULD NOT BE USED TO INITIALIZE A PHYSICAL_OBJECT.
	 * DEFINE DIMENSIONS, AND THE VOLUME WILL BE COMPUTED.
	 * @param volume the new volume
	 * @param updateDiameter if true, the diameter will be updated.
	 */
	void setVolume(double volume, boolean updateDiameter);

	/**
	 * helper function that only sets the volume attribute to the specified value
	 * @param volume
	 */
	void setVolumeOnly(double volume);

	/**
	 * Sets the volume, and recomputes an new diameter.
	 * THIS METHOD SHOULD NOT BE USED TO INITIALIZE A PHYSICAL_OBJECT.
	 * DEFINE DIMENSIONS, AND THE VOLUME WILL BE COMPUTED.
	 * @param volume the new volume
	 */
	void setVolume(double volume);

	/** Returns the length of a cylinder, or the diameter of a sphere.*/
	double getLength();

	/** Get an intracellular and membrane-bound chemicals that are present
	 *  in this PhysicalNode. */
	IntracellularSubstance getIntracellularSubstance(String id);

	/** Add an intracellular or membrane-bound chemicals
	 *  in this PhysicalNode. */
	void addIntracellularSubstance(IntracellularSubstance is);

	/** Remove an intracellular or membrane-bound chemicals that are present
	 *  in this PhysicalNode. */
	void removeIntracellularSubstance(IntracellularSubstance is);

	/** All the intracellular and membrane-bound chemicals that are present
	 *  in this PhysicalNode. */
	Hashtable<String, IntracellularSubstance> getIntracellularSubstances();

	/** All the intracellular and membrane-bound chemicals that are present
	 *  in this PhysicalNode. */
//	void setIntracellularSubstances(
//			Hashtable<String, IntracellularSubstance> intracellularSubstances);

	double getInterObjectForceCoefficient();

	void setInterObjectForceCoefficient(double interObjectForceCoefficient);

	/**
	 * Convenient way to change family links in the neuron tree structure
	 * (mother, daughter branch). This method is useful during elongation and
	 * retraction for intercalation/removal of elements.
	 * of elements.
	 */
	void updateRelative(ini.cx3d.physics.interfaces.PhysicalObject oldRelative, ini.cx3d.physics.interfaces.PhysicalObject newRelative);

	/**
	 * Returns the absolute coordinates of the location where a <code>PhysicalObject</code> is attached
	 * to this <code>PhysicalObject</code>. Does not necessarily contain a check of the identity of the
	 *  element that makes the request.
	 *
	 * @param daughterWhoAsks the PhysicalObject attached to us.
	 * @return the coord
	 */
	double[] originOf(ini.cx3d.physics.interfaces.PhysicalObject daughterWhoAsks);

	/**
	 * Removal of a <code>PhysicalObject</code> from the list of our daughters.
	 * (Mainly in case of complete retraction of the daughter.*/
	void removeDaugther(ini.cx3d.physics.interfaces.PhysicalObject daughterToRemove);

	/**
	 * Resets some computational and physical properties (like the tension, volume) after
	 * a displacement
	 */
	void updateDependentPhysicalVariables();

	/* Diffusion of diffusible IntracellularSubstances between two PhysicalObjects.
	 */
	void diffuseWithThisPhysicalObjects(PhysicalObject po, double distance);

	/* Returns the INSTANCE of IntracellularSubstance in this PhysicalObject with the same id
	 * than the IntracellularSubstance given as argument. If there is no such instance, a
	 * new one with similar properties is created, inserted into the intracellularSubstances
	 * vector and then returned. Only used between subclasses of physicalObject for intracellular
	 * diffusion. C.f. very similar method : PhysicalNode.giveYourSubstanceInstance.
	 */
	IntracellularSubstance giveYouIntracellularSubstanceInstance(IntracellularSubstance templateS);

	/**
	 * Returns the inter-object force that the <code>PhysicalObject</code> in which the method is called applies
	 * onto the <code>PhysicalCylinder</code> given as argument.
	 * @param c
	 * @return
	 */
	double[] getForceOn(PhysicalCylinder c);


	void setTotalForceLastTimeStep(double[] totalForceLastTimeStep);

}

File: src/physics/physical_object.cc
#include "physics/physical_object.h"

#include <cmath>

#include "param.h"
#include "matrix.h"
#include "sim_state_serialization_util.h"
#include "physics/physical_bond.h"
#include "physics/physical_sphere.h"
#include "physics/physical_cylinder.h"
#include "physics/inter_object_force.h"
#include "physics/intracellular_substance.h"
#include "spatial_organization/space_node.h"
#include "synapse/excrescence.h"

namespace cx3d {
namespace physics {

InterObjectForce::UPtr PhysicalObject::inter_object_force_ { nullptr };

InterObjectForce* PhysicalObject::getInterObjectForce() {
  return inter_object_force_.get();
}

void PhysicalObject::setInterObjectForce(InterObjectForce::UPtr force) {
  inter_object_force_ = std::move(force);
}

PhysicalObject::PhysicalObject() {
}

PhysicalObject::~PhysicalObject() {
}

StringBuilder& PhysicalObject::simStateToJson(StringBuilder& sb) const {
  PhysicalNode::simStateToJson(sb);
  SimStateSerializationUtil::removeLastChar(sb);
  sb.append(",");

  SimStateSerializationUtil::keyValue(sb, "interObjectForce", inter_object_force_.get());
  SimStateSerializationUtil::keyValue(sb, "stillExisting", still_existing_);
  SimStateSerializationUtil::keyValue(sb, "onTheSchedulerListForPhysicalObjects",
                                      on_scheduler_list_for_physical_objects_);
  SimStateSerializationUtil::keyValue(sb, "massLocation", mass_location_);
  SimStateSerializationUtil::keyValue(sb, "xAxis", x_axis_);
  SimStateSerializationUtil::keyValue(sb, "yAxis", y_axis_);
  SimStateSerializationUtil::keyValue(sb, "zAxis", z_axis_);
  SimStateSerializationUtil::keyValue(sb, "adherence", adherence_);
  SimStateSerializationUtil::keyValue(sb, "mass", mass_);
  SimStateSerializationUtil::keyValue(sb, "diameter", diameter_);
  SimStateSerializationUtil::keyValue(sb, "volume", volume_);
  SimStateSerializationUtil::keyValue(sb, "color", SimStateSerializationUtil::colorToHexString(color_.getValue()),
                                      true);
  SimStateSerializationUtil::keyValue(sb, "totalForceLastTimeStep", total_force_last_time_step_);
  SimStateSerializationUtil::map(sb, "intracellularSubstances", intracellular_substances_);
  SimStateSerializationUtil::unorderedCollection(sb, "physicalBonds", physical_bonds_);
  SimStateSerializationUtil::unorderedCollection(sb, "excrescences", excrescences_);

  return sb;
}

bool PhysicalObject::isAPhysicalObject() const {
  return true;
}

void PhysicalObject::addExcrescence(Excrescence::UPtr ex) {
  excrescences_.push_back(std::move(ex));
}

void PhysicalObject::removeExcrescence(Excrescence* ex) {
  STLUtil::vectorRemove(excrescences_, ex);
}

bool PhysicalObject::isInContact(PhysicalObject* o) {
  if (o->isAPhysicalSphere()) {
    return isInContactWithSphere(static_cast<PhysicalSphere*>(o));
  } else {
    return isInContactWithCylinder(static_cast<PhysicalCylinder*>(o));
  }
}

std::list<PhysicalObject*> PhysicalObject::getPhysicalObjectsInContact() {
  std::list<PhysicalObject*> po;
  for (auto n : getSoNode()->getNeighbors()) {
    if (n->isAPhysicalObject() && isInContact(static_cast<PhysicalObject*>(n))) {
      po.push_back(static_cast<PhysicalObject*>(n));
    }
  }
  return po;
}

std::array<double, 3> PhysicalObject::transformCoordinatesGlobalToLocal(
    const std::array<double, 3>& positionInGlobalCoord) const {
  return std::array<double, 3> { Matrix::dot(positionInGlobalCoord, x_axis_), Matrix::dot(positionInGlobalCoord,
                                                                                          y_axis_), Matrix::dot(
      positionInGlobalCoord, z_axis_) };
}

std::array<double, 3> PhysicalObject::transformCoordinatesLocalToGlobal(
    const std::array<double, 3>& positionInLocalCoord) const {
  return std::array<double, 3> { positionInLocalCoord[0] * x_axis_[0] + positionInLocalCoord[1] * y_axis_[0]
      + positionInLocalCoord[2] * z_axis_[0], positionInLocalCoord[0] * x_axis_[1]
      + positionInLocalCoord[1] * y_axis_[1] + positionInLocalCoord[2] * z_axis_[1], positionInLocalCoord[0]
      * x_axis_[2] + positionInLocalCoord[1] * y_axis_[2] + positionInLocalCoord[2] * z_axis_[2] };
}

void PhysicalObject::addPhysicalBond(const std::shared_ptr<PhysicalBond>& bond) {
  physical_bonds_.push_back(bond);
}

void PhysicalObject::removePhysicalBond(const std::shared_ptr<PhysicalBond>& bond) {
  physical_bonds_.remove(bond);
}

bool PhysicalObject::getHasAPhysicalBondWith(PhysicalObject* po) {
  for (auto pb : physical_bonds_) {
    if (po == pb->getOppositePhysicalObject(this)) {
      return true;
    }
  }
  return false;
}

std::shared_ptr<PhysicalBond> PhysicalObject::makePhysicalBondWith(PhysicalObject* po) {
  return PhysicalBond::create(this, po);
}

bool PhysicalObject::removePhysicalBondWith(PhysicalObject* po, bool removeThemAll) {
  bool there_was_a_bond = false;
  std::size_t i = 0;
  auto it = physical_bonds_.begin();
  while (it != physical_bonds_.end()) {
    auto pb = *it;
    if (po == pb->getOppositePhysicalObject(this)) {
      physical_bonds_.remove(*it);
      po->physical_bonds_.remove(pb);
      if (!removeThemAll) {
        return true;
      } else {
        there_was_a_bond = true;
        i--;  // we continue to check, and since we removed the ith
      }
    }
    i++;
    it++;
  }
  return there_was_a_bond;
}

double PhysicalObject::getIntracellularConcentration(const std::string& substanceId) {
  if (!STLUtil::mapContains(intracellular_substances_, substanceId)) {
    return 0;
  } else {
    return intracellular_substances_[substanceId]->getConcentration();
  }
}

void PhysicalObject::modifyIntracellularQuantity(const std::string& id, double quantityPerTime) {

  IntracellularSubstance* s = nullptr;
  if (!STLUtil::mapContains(intracellular_substances_, id)) {
    auto s_uptr = PhysicalNode::ecm_->intracellularSubstanceInstance(id);
    s = s_uptr.get();
    intracellular_substances_[id] = std::move(s_uptr);
  } else {
    s = intracellular_substances_[id].get();
  }
  double delta_q = quantityPerTime * Param::kSimulationTimeStep;
  s->changeQuantityFrom(delta_q);
  if (s->isVolumeDependant()) {
    s->updateConcentrationBasedOnQuantity(volume_);
  } else {
    s->updateConcentrationBasedOnQuantity(getLength());
  }
}

double PhysicalObject::getMembraneConcentration(const std::string& id) {
  if (id == "U") {
    return 1.0;
  }
  // otherwise : do we have it on board ?
  if (!STLUtil::mapContains(intracellular_substances_, id)) {
    return 0.0;
  } else {
    // if yes, is it a membrane substance ?
    auto s = intracellular_substances_[id].get();
    if (!s->isVisibleFromOutside()) {
      return 0.0;
    }
    return s->getConcentration();
  }
}

void PhysicalObject::modifyMembraneQuantity(const std::string& id, double quantityPerTime) {
  // for now, the intracellular and membrane bound Substances are the same.
  modifyIntracellularQuantity(id, quantityPerTime);
}

IntracellularSubstance* PhysicalObject::giveYouIntracellularSubstanceInstance(IntracellularSubstance* templateS) {
  IntracellularSubstance* s = nullptr;
  if (!STLUtil::mapContains(intracellular_substances_, templateS->getId())) {
    auto s_uptr = IntracellularSubstance::UPtr(new IntracellularSubstance(*templateS));
    s = s_uptr.get();
    intracellular_substances_[s->getId()] = std::move(s_uptr);
  } else {
    s = intracellular_substances_[templateS->getId()].get();
  }
  return s;
}

void PhysicalObject::diffuseWithThisPhysicalObjects(PhysicalObject* po, double distance) {
  // We store these temporary variable, because we still don't know if the
  // Substances depend on volumes or not

  double vA_v = volume_;
  double vB_v = po->getVolume();
  double pre_a_v = (1.0 / distance);
  double pre_m_v = (1.0 / distance) * (1.0 / vA_v + 1.0 / vB_v);
  double vA_l = getLength();
  double vB_l = po->getLength();
  double pre_a_l = (1.0 / distance);
  double pre_m_l = (1.0 / distance) * (1.0 / vA_l + 1.0 / vB_l);

  // the variable we are effectively using
  double vA;
  double vB;
  double pre_a;
  double pre_m;

  for (auto& element : intracellular_substances_) {
    // for a given substance
    auto s_a = element.second.get();

    // does the substance depend on volumes or not ?
    if (s_a->isVolumeDependant()) {
      vA = vA_v;
      vB = vB_v;
      pre_a = pre_a_v;
      pre_m = pre_m_v;
    } else {
      vA = vA_l;
      vB = vB_l;
      pre_a = pre_a_l;
      pre_m = pre_m_l;
    }

    double s_a_concentration = s_a->getConcentration();

    // stop here if 1) non diffusible substance or 2) concentration very low:
    double diffusion_constant = s_a->getDiffusionConstant();
    if (diffusion_constant < 10E-14 || s_a_concentration < Param::kMinimalConcentrationForIntracellularDiffusion) {
      continue;  // to avoid a division by zero in the n/m if the diff const = 0;
    }

    // find the counterpart in po
    auto s_b = po->giveYouIntracellularSubstanceInstance(s_a);
    double s_b_concentration = s_b->getConcentration();

    // saving time : no diffusion if almost no difference;
    double abs_diff = std::abs(s_a_concentration - s_b_concentration);
    if ((abs_diff < Param::kMinimalConcentrationForIntracellularDiffusion)
        || (abs_diff / s_a_concentration < Param::kMinimalDCOverCForIntracellularDiffusion)) {
      continue;
    }
    // TODO : if needed, when we come here, we have to re-put ourselves on the
    // scheduler list for intra-cellular diffusion.

    // analytic computation of the diffusion between these two PhysicalObjects
    // (cf document "Diffusion" by F.Zubler for explanation).
    double q_a = s_a->getQuantity();
    double q_b = s_b->getQuantity();
    double Tot = q_a + q_b;
    double a = pre_a * diffusion_constant;
    double m = pre_m * diffusion_constant;
    double n = a * Tot / vB;
    double n_over_m = n / m;
    double K = q_a - n_over_m;
    q_a = K * MathUtil::exp(-m * Param::kSimulationTimeStep) + n_over_m;
    q_b = Tot - q_a;

    s_a->setQuantity(q_a);
    s_b->setQuantity(q_b);
    // and update their concentration again
    // (Recall : if volumeDependant = false, vA and vB are not the true volume,
    // but rather the length of the physical object)
    s_a->updateConcentrationBasedOnQuantity(vA);
    s_b->updateConcentrationBasedOnQuantity(vB);
  }
}

Color PhysicalObject::getColor() const {
  return color_;
}

void PhysicalObject::setColor(Color c) {
  color_ = c;
}

std::array<double, 3> PhysicalObject::getMassLocation() const {
  return std::array<double, 3> { mass_location_[0], mass_location_[1], mass_location_[2] };  //todo copy really needed?
}

void PhysicalObject::setMassLocation(const std::array<double, 3>& mass_location) {
  mass_location_ = mass_location;
}

std::array<double, 3> PhysicalObject::getXAxis() const {
  return x_axis_;
}

void PhysicalObject::setXAxis(const std::array<double, 3>& axis) {
  x_axis_ = axis;
}

std::array<double, 3> PhysicalObject::getYAxis() const {
  return y_axis_;
}

void PhysicalObject::setYAxis(const std::array<double, 3>& axis) {
  y_axis_ = axis;
}

std::array<double, 3> PhysicalObject::getZAxis() const {
  return z_axis_;
}

void PhysicalObject::setZAxis(const std::array<double, 3>& axis) {
  z_axis_ = axis;
}

std::array<double, 4> PhysicalObject::getTotalForceLastTimeStep() const {
  return total_force_last_time_step_;
}

bool PhysicalObject::isStillExisting() const {
  return still_existing_;
}

void PhysicalObject::setStillExisting(bool stillExists) {
  still_existing_ = stillExists;
}

bool PhysicalObject::isOnTheSchedulerListForPhysicalObjects() const {
  return on_scheduler_list_for_physical_objects_;
}

void PhysicalObject::setOnTheSchedulerListForPhysicalObjects(bool on_scheduler_list) {
  on_scheduler_list_for_physical_objects_ = on_scheduler_list;
}

std::list<std::shared_ptr<PhysicalBond>> PhysicalObject::getPhysicalBonds() const {
  std::list < std::shared_ptr<PhysicalBond> > list;
  for (auto pb : physical_bonds_) {
    list.push_back(pb);
  }
  return list;
}

void PhysicalObject::setPhysicalBonds(const std::list<std::shared_ptr<PhysicalBond> >& bonds) {  //todo change to vector
  physical_bonds_ = bonds;
}

std::vector<Excrescence*> PhysicalObject::getExcrescences() const {
  std::vector<Excrescence*> ret;
  for (auto& ex : excrescences_) {
    ret.push_back(ex.get());
  }
  return ret;
}

double PhysicalObject::getAdherence() const {
  return adherence_;
}

void PhysicalObject::setAdherence(double a) {
  adherence_ = a;
}

double PhysicalObject::getMass() const {
  return mass_;
}

void PhysicalObject::setMass(double m) {
  mass_ = m;
}

double PhysicalObject::getDiameter() const {
  return diameter_;
}

void PhysicalObject::setDiameter(double diameter) {
  setDiameter(diameter, true);
}

void PhysicalObject::setDiameter(double d, bool update_volume) {
  diameter_ = d;
  if (update_volume) {
    updateVolume();
  }
}

double PhysicalObject::getVolume() const {
  return volume_;
}

void PhysicalObject::setVolume(double v, bool update_diameter) {
  volume_ = v;
  updateIntracellularConcentrations();
  if (update_diameter) {
    updateDiameter();
  }
}

void PhysicalObject::setVolume(double volume) {
  setVolume(volume, true);
}

IntracellularSubstance* PhysicalObject::getIntracellularSubstance(const std::string& id) {
  if (STLUtil::mapContains(intracellular_substances_, id)) {
    return intracellular_substances_[id].get();
  }
  return nullptr;
}

void PhysicalObject::addIntracellularSubstance(IntracellularSubstance::UPtr is) {
  intracellular_substances_[is->getId()] = std::move(is);
}

void PhysicalObject::removeIntracellularSubstance(IntracellularSubstance* is) {
  intracellular_substances_.erase(is->getId());
}

std::list<IntracellularSubstance*> PhysicalObject::getIntracellularSubstances1() const {
  std::list<IntracellularSubstance*> list;
  for (auto& el : intracellular_substances_) {
    list.push_back(el.second.get());
  }
  return list;
}

void PhysicalObject::addNewIntracellularSubstance(IntracellularSubstance::UPtr s) {
  intracellular_substances_[s->getId()] = std::move(s);
}

void PhysicalObject::setVolumeOnly(double v) {
  volume_ = v;
}

}  // namespace physics
}  // namespace cx3d
